OTAnalytics/plugin_video_processing/video_reader.py

```diff
@@ -1,52 +1,78 @@
+    return PilImage(frame.to_image().convert(GRAYSCALE))
 
 
-class FrameDoesNotExistError(Exception):
-    pass
+class PyAvVideoReader(VideoReader):
+    def get_fps(self, video_path: Path) -> float:
+        with self.__get_clip(video_path) as container:
+            rate = self.__get_fps(container, video_path)
+            return rate.numerator / rate.denominator
+        raise ValueError(f"Could not read frames per second from {str(video_path)}")
 
+    def __get_fps(self, container: InputContainer, video_path: Path) -> Fraction:
+        average_rate = container.streams.video[0].average_rate
+        if average_rate is None:
+            raise ValueError(f"Could not read frames per second from {str(video_path)}")
+        return average_rate
 
-class OpenCvVideoReader(VideoReader):
-    def get_fps(self, video_path: Path) -> float:
-        return self.__get_clip(video_path).get(cv2.CAP_PROP_FPS)
+    def get_frame(self, video_path: Path, frame_number: int) -> TrackImage:
+        """Get image of video at position `frame_number`.
+
+        It uses PyAV to seek the closest keyframe. Afterwards, it iterates forward
+        through the video to find the correct frame. Given this implementation,
+        the complexity is O(n).
 
-    def get_frame(self, video_path: Path, index: int) -> TrackImage:
-        """Get image of video at `frame`.
         Args:
             video_path (Path): path to the video_path.
-            index (int): the frame of the video to get.
+            frame_number (int): the frame of the video to get.
         Raises:
             FrameDoesNotExistError: if frame does not exist.
         Returns:
             ndarray: the image as an multi-dimensional array.
         """
-        cap = self.__get_clip(video_path)
-        total_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
-        frame_to_load = min(index, (total_frames - 1))
-        cap.set(cv2.CAP_PROP_POS_FRAMES, frame_to_load)
-        is_read, bgr_frame = cap.read()
-        cap.release()
-        rgb_frame = bgr_frame[:, :, ::-1]
-        return PilImage(Image.fromarray(rgb_frame).convert(GRAYSCALE))
+        frame = self._read_frame(frame_number, video_path)
+        return av_to_image(frame)
+
+    def _read_frame(self, frame_to_read: int, video_path: Path) -> VideoFrame:
+        with self.__get_clip(video_path) as container:
+            if len(container.streams.video) <= 0:
+                raise InvalidVideoError(f"{str(video_path)} is not a video")
+            video = container.streams.video[0]
+            max_frames = video.frames
+            frame_to_read = min(frame_to_read, max_frames - OFFSET)
+            framerate = self.__get_fps(container, video_path)
+            time_base = (
+                video.time_base if video.time_base else Fraction(av.time_base, 1)
+            )
+            time_in_video = int(frame_to_read / framerate)
+            container.seek(time_in_video * av.time_base, backward=True)
```

Comment:
Is there a reason why we are not using time_base that we declared above?

Reply:
Yes, I refactored the code and added some documentation.